drivers/sensor/ina219/ina219.c

```diff
@@ -0,0 +1,381 @@
+			&reg_addr, sizeof(reg_addr),
+			rx_buf, sizeof(rx_buf));
+
+	*reg_data = sys_get_be16(rx_buf);
+
+	return rc;
+}
+
+static int ina219_reg_write(const struct device *dev,
+		uint8_t addr,
+		uint16_t reg_data)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint8_t tx_buf[3];
+
+	tx_buf[0] = addr;
+	sys_put_be16(reg_data, &tx_buf[1]);
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf), cfg->i2c_addr);
+}
+
+static int ina219_reg_field_update(const struct device *dev,
+		uint8_t addr,
+		uint16_t mask,
+		uint16_t field)
+{
+	uint16_t reg_data;
+	int rc;
+
+	rc = ina219_reg_read(dev, addr, &reg_data);
+	if (rc) {
+		return rc;
+	}
+
+	reg_data = (reg_data & ~mask) | field;
+
+	return ina219_reg_write(dev, addr, reg_data);
+}
+
+static int ina219_set_msr_delay(const struct device *dev)
+{
+	const struct ina219_config *cfg = dev->config;
+	struct ina219_data *data = dev->data;
+
+	data->msr_delay = ina219_conv_delay(cfg->badc) +
+		ina219_conv_delay(cfg->sadc);
+	return 0;
+}
+
+static int ina219_set_config(const struct device *dev)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint16_t reg_data;
+
+	reg_data = (cfg->brng & INA219_BRNG_MASK) << INA219_BRNG_SHIFT |
+		(cfg->pg & INA219_PG_MASK) << INA219_PG_SHIFT |
+		(cfg->badc & INA219_ADC_MASK) << INA219_BADC_SHIFT |
+		(cfg->sadc & INA219_ADC_MASK) << INA219_SADC_SHIFT |
+		(cfg->mode & INA219_MODE_NORMAL);
+
+
+	return ina219_reg_write(dev, INA219_REG_CONF, reg_data);
+}
+
+static int ina219_set_calib(const struct device *dev)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint16_t cal;
+
+	cal = INA219_SCALING_FACTOR / ((cfg->r_shunt) * (cfg->current_lsb));
+
+	return ina219_reg_write(dev, INA219_REG_CALIB, cal);
+}
+
+static int ina219_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct ina219_data *data = dev->data;
+	uint16_t status;
+	uint16_t tmp;
+	int rc;
+	enum sensor_channel_ina219 chan_p = (enum sensor_channel_ina219)chan;
+
+	if (chan != SENSOR_CHAN_ALL &&
+		chan_p != SENSOR_CHAN_INA219_V_SHUNT &&
+		chan_p != SENSOR_CHAN_INA219_BUS_VOLTAGE &&
+		chan_p != SENSOR_CHAN_INA219_BUS_POWER &&
+		chan_p != SENSOR_CHAN_INA219_BUS_CURRENT) {
+		return -ENOTSUP;
+	}
+
+	/* Trigger measurement and wait for completion */
+	rc = ina219_reg_field_update(dev,
+				    INA219_REG_CONF,
+				    INA219_MODE_MASK,
+				    INA219_MODE_NORMAL);
+	if (rc) {
+		LOG_ERR("Failed to start measurement.");
+		return rc;
+	}
+
+	k_sleep(K_USEC(data->msr_delay));
+
+	rc = ina219_reg_read(dev, INA219_REG_V_BUS, &status);
+	if (rc) {
+		LOG_ERR("Failed to read device status.");
+		return rc;
+	}
+
+	while (!(INA219_CNVR_RDY(status))) {
+		rc = ina219_reg_read(dev, INA219_REG_V_BUS, &status);
+		if (rc) {
+			LOG_ERR("Failed to read device status.");
+			return rc;
+		}
+		k_sleep(K_USEC(INA219_WAIT_MSR_RETRY));
+	}
+
+	/* Check for overflow */
+	if (INA219_OVF_STATUS(status)) {
+		LOG_ERR("Power and/or Current calculations are out of range.");
```

Comment:
If it's an acceptable case, then I'd change to `LOG_WRN`

Reply:
i did just that. i think it is fine this way.
thanks for all the reviewing